File: 01_SOLID/05_DependencyInversionPrinciple (DIP)/cpp/badCode.cpp
#include <iostream>
#include <memory>

using namespace std;

// Class implementing the recommendations based on recently added
class RecentlyAdded {
public:
    void getRecommendations() {
        cout << "Showing recently added content..." << endl;
    }
};

class TrendingNow {
public:
    void getTrendingRecommendations() {
        cout << "Showing trending content..." << endl;
    }
};

class GenreBased {
public:
    void getGenreRecommendations() {
        cout << "Showing content based on your favorite genres..." << endl;
    }
};

// Class implementing the overall Recommendation Engine
class RecommendationEngine {
private:
    RecentlyAdded recommender;
    TrendingNow trendingRecommender;
    GenreBased genreRecommender;

public:
    void recommend() {
        recommender.getRecommendations();
    }
    
    void recommendTrending() {
        trendingRecommender.getTrendingRecommendations();
    }
    
    void recommendGenre() {
        genreRecommender.getGenreRecommendations();
    }
};

int main() {
    RecommendationEngine engine;
    engine.recommend();
    engine.recommendTrending();
    engine.recommendGenre();
    return 0;
}

File: 01_SOLID/05_DependencyInversionPrinciple (DIP)/badCode.java
// Class implementing the recommendations based on recently added
class RecentlyAdded {
    // Method to get the recommendations
    public void getRecommendations() {
        System.out.println("Showing recently added content...");
    }
}

class TrendingNow {
    // Method to get the recommendations
    public void getTrendingRecommendations() {
        System.out.println("Showing trending content...");
    }
}

class GenreBased {
    // Method to get the recommendations
    public void getGenraRecommendations() {
        System.out.println("Showing content based on your favorite genres...");
    }
}

// Class implementing the overall Recommendation Engine
class RecommendationEngine {
    private RecentlyAdded recommender = new RecentlyAdded();
    private TrendingNow trendingRecommender = new TrendingNow();
    private GenreBased genreRecommender = new GenreBased();

    public void recommend() {
        recommender.getRecommendations();
    }
    public void recommendTrending() {
        trendingRecommender.getTrendingRecommendations();
    }
    public void recommendGenre() {
        genreRecommender.getGenraRecommendations();
    }
    

}
